### PR TITLE
Support debuginfo for custom MIR.

### PR DESCRIPTION
**Pull request by cjgillot**

### CODE REVIEW

**Comment by petrochenkov** [comment]
r? compiler

**Comment by RalfJung** [comment]
Did this break stage 0 Miri tests?
```
tests/pass/function_calls/return_place_on_heap.rs FAILED:
command: "/home/r/src/rust/rustc/build/x86_64-unknown-linux-gnu/stage0-tools-bin/miri" "--error-format=json" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--target" "x86_64-unknown-linux-gnu" "--out-dir" "/home/r/src/rust/rustc/build/x86_64-unknown-linux-gnu/stage0-tools/ui" "tests/pass/function_calls/return_place_on_heap.rs" "--edition" "2021"




pass test got exit status: 1, but expected 0

actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/function_calls/return_place_on_heap.stderr` to the actual output
--- tests/pass/function_calls/return_place_on_heap.stderr
+++ <stderr output>
+error: Could not parse debuginfo, found: "Block { block: b0 }"
+  --> $DIR/return_place_on_heap.rs:LL:CC
+   |
+LL | /     mir! {
+LL | |         {
+LL | |             let x = 0;
+LL | |             let ptr = &raw mut x;
+...  |
+LL | |         }
+LL | |     }
+   | |_____^
+   |
+   = note: this error originates in the macro `::core::intrinsics::mir::__internal_remove_let` which comes from the expansion of the macro `mir` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error: aborting due to previous error
+
+


There were 1 unmatched diagnostics at tests/pass/function_calls/return_place_on_heap.rs:10
    Error: Could not parse debuginfo, found: "Block { block: b0 }"

full stderr:
error: Could not parse debuginfo, found: "Block { block: b0 }"
  --> tests/pass/function_calls/return_place_on_heap.rs:10:5
   |
LL | /     mir! {
LL | |         {
LL | |             let x = 0;
LL | |             let ptr = &raw mut x;
...  |
LL | |         }
LL | |     }
   | |_____^
   |
   = note: this error originates in the macro `::core::intrinsics::mir::__internal_remove_let` which comes from the expansion of the macro `mir` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error



FAILURES:
    tests/pass/align_repeat_into_packed_field.rs
    tests/pass/unsized.rs (revision stack)
    tests/pass/unsized.rs (revision tree)
    tests/pass/function_calls/return_place_on_heap.rs
```
On CI we only run stage 2 tests, but for local development stage 0 tests are very valuable.

File: library/core/src/intrinsics/mir.rs [hunk 1]
```diff
@@ -371,6 +379,7 @@ pub macro mir {
     (
         $(type RET = $ret_ty:ty ;)?
         $(let $local_decl:ident $(: $local_decl_ty:ty)? ;)*
+        $(debug $dbg_name:ident => $dbg_data:expr ;)*
```
Comment (RalfJung) [inline library/core/src/intrinsics/mir.rs:382]: Would be great if some doc comment somewhere explained this syntax.

**Pull request merged**